Config setting to limit deep pagination for results

lahmann opened the pull request:
Demian, this is a fairly simple implementation of a setting in `searches.ini` to limit maximum pages for results in order to avoid high loads on Solr due to deep pagination. Although this might be handled differently with [Solr 4.7+](https://issues.apache.org/jira/browse/SOLR-5463) this implementation can be helpful for shared-index setups. Especially if the catalogues are being crawled this can cause very high loads plus it's doubtable that users are likely to decent deeper than 20 pages of results. E.g. [Google does this too](https://www.google.com/?gfe_rd=cr&ei=AxemVdiEAsaK8QfJ6bCgCg&gws_rd=cr#q=vufind&start=1000), so it is at least a common practise.

demiankatz commented:
André,

We actually already have a feature like this for Summon because the API sets a hard limit on number of results. This may be implemented for some other backends as well (though I can't remember off the top of my head). You can see the setting here:

https://github.com/vufind-org/vufind/blob/master/config/vufind/Summon.ini#L77

Because of the specifics of the Summon API, we were forced to limit to a total maximum number of results rather than a total number of pages (since page size can be configured to be variable, and we don't want to allow users to get deeper into the set simply by making the page size bigger). I think this is probably a best practice across the board because it's more consistent when page size is variable.

Rather than implementing two different mechanisms, I think it would probably make more sense to make the Solr limit consistent with the Summon one (and any others that already exist). Perhaps this could either be bumped into the base class or turned into a Trait to reduce redundancy.

What do you think? (And apologies if there are details in your code that are not covered by the Summon implementation I mentioned -- I haven't looked at it in detail, since I figured it was better to have this discussion before reviewing anything).

lahmann commented:
Demian, thanks a lot for pointing us to this Summon-specific feature. We will definitely look into the Summon-Logic and will try to adapt the code. After taking a brief look at it I don't think that our code covers other details than the Summon implementation. When we've made progress we'll post it in this pull request, or do you prefer opening a new pull request?

demiankatz commented:
Posting it here is fine -- that keeps the continuity of our discussion, and I can just flatten the commit history when it comes time to merge to master. Thanks!

lahmann commented:
We removed all the code from the previous commits and adjusted `VuFind\Search\Base\Options->getVisibleSearchResultLimit()` so that the `result_limit` setting from searches.ini will be used if set - otherwise the default value `-1` will be returned.
Limiting this to Solr would have been an option, but would have meant to introduce a separate Solr.ini - if I am correct.

What's not addressed is altering the get-parameters: the `result_limit` setting does not block manipulating the `page`-parameter. So that even Summon-results can be extended: e.g. http://demovufind.cineca.it/vufind/Summon/Search?type=AllFields&page=50 (until Summon caps the VuFind-request which appears to be happen above 50 pages). 

We addressed this issue in `VuFind\Search\Base\Results->getEndRecord()` and `VuFind\Search\Base\Results->getStartRecord()` (as you can see in 98a210c4), but maybe there is a better way to implement this?

demiankatz commented:
Regarding a separate Solr.ini, you could actually consider searches.ini to be Solr.ini -- it only contains Solr-specific settings. Either way, though, I'm fine with putting this capability into the base. If you make the modifications you propose to getStartRecord and getEndRecord in base, that should universally restrict lookups with the setting, so it should be globally supported (unless one of the backends is doing something unconventional... which hopefully is not the case).

It's probably worth searching the code for other implementations of getVisibleSearchResultLimit and removing them if they are now covered by the base class. (I can test these for you if there's any question).

Good point regarding Summon allowing circumvention via the page parameter. At the time, I wasn't too concerned about this, because the setting was intended to prevent users from encountering Summon error messages rather than preventing them from seeing deep results... but it does make sense for the sake of consistency to enforce this better.

Thanks again!

demiankatz commented:
This looked like it could be simplified, so rather than trying to discuss it, I just took a stab at simplification (I also unified this with the existing Primo/Summon/EDS implementations). Please take a look and let me know if you notice any problems.

There is one known issue: if you set the limit to a number that is not evenly divisible by the page limit, the last page will include some extra results in Solr (i.e. if you set the limit to 98, you'll still see results 99 and 100). Overcoming this is probably more complicated than it's worth, though I'm sure it could be done if it were considered important enough.

gawol commented:
Demian,

everything is okay, it works. I found no problems.

lahmann commented:
Thanks a lot Demian! As Gregor already stated, the code works fine for us. So if we don't want to address the limit-pagelimit-division issue (which I think is not too important), we should be done here.

demiankatz commented:
Thanks for the checks. I've manually merged this to master (to simplify the commit history).
